Add code generator for Go

andrewkroh opened the pull request:
This generates Go code based on the Elastic Common Schema.

I find myself creating structs that match the schema so I thought why not generate some code.

ruflin commented:
Very interesting idea. We could do that for other languages too. I assume you plan to import the code directly from ECS so if it gets updated and you go to a new version, you also get the newest fields?

andrewkroh commented:
@ruflin I updated the PR to include two README files. I also added

- version.go with a `const Version`
- doc.go to make the godocs better.
- Added a `VERSION` variable to the Makefile to make updating the version a little easier.

andrewkroh commented on code/go/ecs/agent.go:
```diff
@@ -0,0 +1,50 @@
+	// readable separation is needed on which Filebeat instance data is coming
+	// from.
+	// If no name is given, the name is often left empty.
+	Name string `ecs:"name"`
```
I haven't made use of it yet, but it's there to, at minimum, make it possible to marshal/unmarshal these fields in a programatic manner (if you take the time to read the tag values). I can see a future enhancements to make this better.

webmat commented:
Thanks for sumitting this! I'll look after the Beta 2 release, however ;-)

ruflin commented on Makefile:
```diff
@@ -1,12 +1,14 @@
-generate: schemas readme template fields
+VERSION ?= 1.0.0-beta1
```
@andrewkroh Should be perhaps introduce a `.version` file in the repo where all tools can read from?

ruflin commented on Makefile:
```diff
@@ -47,4 +49,7 @@ fields:
 	rm -f fields.tmp.yml fields.tmp.yml.bak
 
-.PHONY: generate schemas fmt check setup clean readme template fields
+codegen:
```
We should probably also move to `mage` on day ;-)

ruflin commented on scripts/template.go:
```diff
@@ -1,22 +1,55 @@
+	log.SetFlags(0)
+	flag.Parse()
+
+	if version == "" {
```
Perhaps we the change around the version for the template and code cleanup to a separate PR to get in quickly?

andrewkroh replied:
I was thinking that when I made this change, but then I figured it wasn’t a very important change so it could wait. But if you’d like I can pull this one out into its own PR?

ruflin replied:
Would be great to have it seperate. I think it's important as it changes the template a bit.

andrewkroh replied:
Created #279.

ph commented:
@urso I think it would be a good time to show your logger experiment because I believe its kinda linked to this generated code :)

urso commented on code/go/ecs/agent.go:
```diff
@@ -0,0 +1,52 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Code generated by scripts/gocodegen.go - DO NOT EDIT.
```
Not sure if the format is correct, but If possible we should have this at the top of the file. This way github will also hide the diff by default.

urso commented:
It would be nice if we could define a template to set how the output should look like. E.g. in my logging experiments I did come up with:

```
	log.With(
		ecs.Service.Name("my server"),
		ecs.Hostname("localhost"),
	).Info("start service")
```

or:

```
// pass to first subsystem
log = log.With(ecs.Service.Name("my server"))

// second subsystem
log = log.With(ecs.Destination.Port(8080))

// actual go routine handling an http request...

log = log.With(
  	ecs.HTTP.Method("GET"),
	ecs.HTTP.URL("/get_file/file.txt"),
	ecs.Source.Domain("localhost"),
	ecs.Source.IP("127.0.0.1"),
)
...
log.Info("can not read {filename}", "file.txt") // filename becomes an event field {fields.filename: "file.txt"}
```

One can add a field to an existing namespace and everything is typed + deduplicated when logging. More similar how zap/zerolog and other structured loggers work. The structures I create for this are like:

```
type serviceNS struct{}

var Service serviceNS

func (serviceNS) Name(name string) Field { return Field{Key: "service.name", Value: name} }
```

Nesting namespaces goes like this:

```
type logNS struct {
  File logFileNS
}
type logFileNS struct{}

...

func (logNS) Level(s string) fld.Field        { return ecsString("log.level", s) }
func (logFileNS) Path(s string) fld.Field     { return ecsString("log.file.path", s) }
```

webmat commented:
As discussed elsewhere, the copyright holder should actually be "Elasticsearch B.V.", not "Elastic N.V."

My other two review feedback still stand, however:

- missing base fields
- separate MIT code from Apache code :-)

andrewkroh commented:
@webmat Both points are addressed now. I'm glad you caught that we were missing the base fields.

webmat commented:
@andrewkroh Looking good! I wonder why you had `device` fields, though. What git ref was this based on?

andrewkroh commented:
`device` was there because it existed when I started, but there wasn't anything to delete the generated .go files until now.

webmat commented:
Good catch on `User_agent`
